app/src/main/java/io/neurolab/program_modes/MemoryGraphActivity.java

```diff
@@ -18,13 +28,71 @@
 import com.github.mikephil.charting.interfaces.datasets.ILineDataSet;
 import com.github.mikephil.charting.listener.OnChartValueSelectedListener;
 
+import java.io.UnsupportedEncodingException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
 import io.neurolab.R;
+import io.neurolab.communication.CommunicationHandler;
 import io.neurolab.main.MainActivity;
 
 public class MemoryGraphActivity extends AppCompatActivity implements OnChartValueSelectedListener {
 
+    private final String ACTION_USB_PERMISSION = "io.neurolab.USB_PERMISSION";
     private LineChart memGraph;
     private Thread thread;
+    private UsbManager usbManager;
+    private UsbSerialDevice serialPort;
+    private CommunicationHandler communicationHandler;
+    private ArrayList eegDataValues;
+
+    private UsbSerialInterface.UsbReadCallback readCallback = new UsbSerialInterface.UsbReadCallback() { //Defining a Callback which triggers whenever data is read.
+        @Override
+        public void onReceivedData(byte[] arg0) {
+            String data = null;
+            try {
+                data = new String(arg0, "UTF-8");
+                eegDataValues = parseDataForGraph(data);
+            } catch (UnsupportedEncodingException e) {
+                e.printStackTrace();
+            }
+        }
+    };
+    private final BroadcastReceiver broadcastReceiver = new BroadcastReceiver() { //Broadcast Receiver to automatically start and stop the Serial connection.
+        @Override
```

Comment:
We need to move all connection related functions to entry point activity and let each UI related activities use them without reiterating same code

Reply:
Yes @CloudyPadmal will direct to Launcher asap.